### PR TITLE
Add JSON Feed support

### PR DESCRIPTION
**Pull request by ozkutuk**
> The JSON Feed format is a pragmatic syndication format, like [RSS](http://cyber.harvard.edu/rss/rss.html) and [Atom](https://tools.ietf.org/html/rfc4287), but with one big difference: it’s JSON instead of XML.

This PR adds JSON Feed support to Hakyll with a pretty barebone template, conforming to the [JSON Feed 1.1 spec](https://www.jsonfeed.org/version/1.1/). I have deployed it to my [personal blog](https://ozkutuk.me/feed.json) to serve as a working example, and it seems to work without any issues.

### CODE REVIEW

File: lib/Hakyll/Web/Feed.hs [hunk 1]
```diff
@@ -176,8 +220,32 @@ renderAtom :: FeedConfiguration       -- ^ Feed configuration
...
 -- | Copies @$updated$@ from @$published$@ if it is not already set.
 makeItemContext :: String -> Context a -> Context a
 makeItemContext fmt context = mconcat
     [context, dateField "published" fmt, dateField "updated" fmt]
+
+
+--------------------------------------------------------------------------------
+-- | Escape the string with the usual Haskell escape conventions while leaving
+-- the non-ASCII characters intact.
+escapeString :: String -> String
```
Comment (Minoru) [inline lib/Hakyll/Web/Feed.hs:242]: This function renders a "delete" character as `\DEL`, which is not valid JSON. I think it'd be best to model this after [RFC 8259 §7](https://www.rfc-editor.org/rfc/rfc8259#section-7). Can you please do that?

File: lib/Hakyll/Web/Feed.hs [hunk 2]
```diff
@@ -176,8 +220,32 @@ renderAtom :: FeedConfiguration       -- ^ Feed configuration
 renderAtom = renderAtomWithTemplates atomTemplate atomItemTemplate
 
 
+--------------------------------------------------------------------------------
+-- | Render a JSON feed with a number of items.
+renderJsonFeed :: FeedConfiguration       -- ^ Feed configuration
```
Comment (Minoru) [inline lib/Hakyll/Web/Feed.hs:225]: Other rendering functions don't have "Feed" in their hame, can you please change this one to match?

File: lib/Hakyll/Web/Feed.hs [hunk 3]
```diff
@@ -154,7 +184,21 @@ renderAtomWithTemplates ::
...
+--------------------------------------------------------------------------------
+-- | Render a JSON feed using given templates with a number of items.
+renderJsonFeedWithTemplates ::
+       Template                -- ^ Feed template
+    -> Template                -- ^ Item template
+    -> FeedConfiguration       -- ^ Feed configuration
+    -> Context String          -- ^ Item context
+    -> [Item String]           -- ^ Feed items
+    -> Compiler (Item String)  -- ^ Resulting feed
+renderJsonFeedWithTemplates feedTemplate itemTemplate config context = renderFeed
```
Comment (Minoru) [inline lib/Hakyll/Web/Feed.hs:200]: For consistency with other functions, this should be renamed to `renderJsonWithTemplates` and added to the exports list.

File: data/templates/feed-item.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "id": "$root$$url$",
+    "url": "$root$$url$",
+    "content_html": "$description$",
```
Comment (Minoru) [inline data/templates/feed-item.json:4]: JSON feed offers two fields for the body: `content` and `content_html`. I think it's important to document that the default template shipped with Hakyll uses `content_html`.

**Review by Minoru — COMMENTED** [review]
Thank you, this looks excellent! I left a few nitpicks in comments below, please take a look.

File: lib/Hakyll/Web/Feed.hs (hunk 1 above)
Comment (ozkutuk) [reply lib/Hakyll/Web/Feed.hs:246]: I modified the escaping logic to match the RFC, can you please take another look?

File: data/templates/feed-item.json (hunk 4 above)
Comment (ozkutuk) [reply data/templates/feed-item.json:4]: Where should that be documented? Is the haddock of `renderJson` a good place for it?

Comment (Minoru) [reply data/templates/feed-item.json:4]: Apologies for such a delayed response! Yeah, `renderJson` is the best place for it. To save us both from another review round, I'll update it myself -- I'll have to rebase this PR anyway to apply CI fixes that I'm currently pushing.

Comment (Minoru) [reply data/templates/feed-item.json:4]: Done: 1456f3eda398c9f824c7e569bf7f3f03b4697c1a

**Review by Minoru — APPROVED** [review]
Will merge once the CI is done.

Thank you for the work, @ozkutuk, and for putting up with such delays in reviewing this!

**Pull request merged**